marda_registry/models/__init__.py

```diff
@@ -3,9 +3,52 @@
 be edited by hand.
 """
 
-__version__ = "0.0.1"
+from pydantic import validator
 
-from .extractor import Extractor
+from .extractor import Extractor as AutoGenExtractor
 from .filetype import FileType
 
+
+class Extractor(AutoGenExtractor):
+    """A wrapper for the automatically generated Extractor class
+    that includes some additional validators.
+
+    """
+
+    @validator("usage")
+    def check_usage_filetypes(cls, v, values):
+        """Check that any filetype instructions in the usage
+        are defined at the top-level, and that all supported_filetypes
+        values have instructions (or that there is a catch-all).
+
+        """
+
+        if not v:
+            return v
+
+        supported_filetypes = values.get("supported_filetypes")
+        usage_to_fts = {}
+
+        bad_fts = set()
+        if usage := v:
+            for instruction in usage:
+                if instruction.filetypes is None:
+                    for t in supported_filetypes:
+                        if t not in usage_to_fts:
+                            usage_to_fts[t] = instruction
+                    for t in instruction.filetypes:
+                        usage_to_fts[t] = instruction
+                        if t not in supported_filetypes:
+                            bad_fts.add((t, "Bad filetype {t} found in instruction"))
+
+        for t in supported_filetypes:
+            if t not in usage_to_fts:
+                bad_fts.add((t, f"No instruction for {t}"))
+
+        if bad_fts:
+            raise ValueError("Unsupported filetypes: " + "\n".join(bad_fts))
```

Comment:
Maybe raise a `pydantic.ValidationError` instead, just in case someone downstream would be catching those?

Reply:
Raising `ValueError` from validators is actually intentional, pydantic converts it to `ValidationError` automatically

Reply:
(see e.g. https://docs.pydantic.dev/latest/concepts/validators/#field-validators -- ctrl+f for ValueError as I can't link to the exact paragraph)